### ISSUE TITLE
ISPC codegen backend and floating point values

### ISSUE THREAD

**Issue by pramodk** [BlueBrain/nmodl#442]
ISPC backend needs to print floating point values in different form because floating point values are treated as `float` by default. From [ISPC docs](https://ispc.github.io/ispc.html):

```
Updating ISPC Programs For Changes In ISPC 1.5.0

This release adds support for double precision floating point constants.
Double precision floating point constants are floating point number with
d suffix and optional exponent part. Here are some examples:
3.14d, 31.4d-1, 1.d, 1.0d, 1d-2. Note that floating point number without
suffix is treated as single precision constant.
```

For this reason, ISPC has `double_to_string` routine to convert floating point values into ISPC compatible form:   https://github.com/BlueBrain/nmodl/blob/ac46928a78cba40fa97900835814cfc765b143ff/src/codegen/codegen_ispc_visitor.cpp#L126

But as observed in #431 and #144, this could result in discrepancies in C++ backend and ISPC backend.

As part of this ticket we need to review the implementation and see if values calculated for units is same/similar to CPP backend.

One option could be [hex representation](https://ispc.github.io/ispc.html#lexical-structure): 

```
Floating-point constants can be specified in one of three ways. First, they may be a sequence of zero or
more digits from 0 to 9, followed by a period, followed by zero or more digits from 0 to 9. (There must be
at least one digit before or after the period).

The second option is scientific notation, where a base value is specified as the first form of a floating-point
constant but is then followed by an "e" or "E", then a plus sign or a minus sign, and then an exponent.

Finally, floating-point constants may be specified as hexadecimal constants; this form can ensure a perfectly
bit-accurate representation of a particular floating-point number. These are specified with an "0x" prefix,
followed by a zero or a one, a period, and then the remainder of the mantissa in hexadecimal form, with digits
from 0-9, a-f, or A-F. The start of the exponent is denoted by a "p", which is then followed by an optional plus
or minus sign and then digits from 0 to 9. For example:

float two = 0x1p+1;  // 2.0
float pi  = 0x1.921fb54442d18p+1;  // 3.1415926535...
float neg = -0x1.ffep+11;  // -4095.

Floating-point constants can optionally have a "f" or "F" suffix (ispc currently treats all floating-point
constants as having 32-bit precision, making this suffix not currently have an effect.)
```

**Comment by pramodk** [comment]
Tested current use cases in https://github.com/BlueBrain/nmodl/pull/431#issuecomment-73914887 and they seems to be working fine. We need to identify if there is/will be any corner case.

**Comment by iomaganaris** [comment]
In `ISPC` codegen currently we are printing double values with a certain accuracy.
See: https://github.com/BlueBrain/nmodl/blob/ac46928a78cba40fa97900835814cfc765b143ff/src/codegen/codegen_ispc_visitor.cpp#L131
This is problematic when constants less than `1e-18` are defined. I came across this when compiling this file: https://github.com/iomaganaris/netpyne/blob/977307b5ccdf4aadd0a6a4181feadfd7a98d5255/examples/coreneuron/m1/mod/kBK.mod#L93
and the printed value in the `.ispc` file is `0.0000000000000000000d`.
I think that we should handle the constant in `ISPC` as strings, same as the `c++` codegen.
I believe it will be fairly easy to do this by converting whatever scientific notation is used to define the constant in the first place to an `ISPC` format string.
To do this we need to:
1. Convert `e` and `E` to `d` for double constants
2. Add `d` in the end of a double variable if it's not defined with scientific notation
3. Add `f` in the end of a float variable if it's not defined with scientific notation
Another option would be using hexadecimal values however I am afraid that this might add discrepancies in the generated results compared to the `c++` backend.

I have started some initial work to solve this issue here: https://github.com/BlueBrain/nmodl/tree/magkanar/ispc_string_doubles
